File: KantaroBLL/Scala.vb
Public Class Scala
    Private _NotaFondamentale As Nota
    Private _Modo As ModoBase

    Public Sub New(notaFondamentale As Nota, modo As ModoBase)
        If notaFondamentale Is Nothing Then Throw New ArgumentNullException("notaFondamentale")
        If modo Is Nothing Then Throw New ArgumentNullException("modo")

        _NotaFondamentale = notaFondamentale
        _Modo = modo
    End Sub

    Public ReadOnly Property NotaFondamentale As Nota
        Get
            Return _NotaFondamentale
        End Get
    End Property
    Public ReadOnly Property Modo As ModoBase
        Get
            Return _Modo
        End Get
    End Property

    Default Public ReadOnly Property Item(index As Integer) As Nota
        Get
            Dim incrementi = index Mod Modo.NumeroSuccessioni
            Dim r As Nota = Me.NotaFondamentale
            For i = 0 To incrementi - 1
                r += Modo.Successioni(i)
            Next
            Return r
        End Get
    End Property

#Region " Verifica uguaglianze "
    Public Shared Operator =(ByVal class1 As Scala, ByVal class2 As Scala) As Boolean
        Return class1.Equals(class2)
    End Operator

    Public Shared Operator <>(ByVal class1 As Scala, ByVal class2 As Scala) As Boolean
        Return Not class1.Equals(class2)
    End Operator

    Public Overrides Function Equals(obj As Object) As Boolean
        If Not TypeOf obj Is Scala Then Return False
        Dim s = DirectCast(obj, Scala)
        Dim bNota = s.NotaFondamentale.Equals(Me.NotaFondamentale)
        Dim bModo = s.Modo.Equals(s.Modo)
        Return bNota AndAlso bModo
    End Function

    Public Overloads Shared Function Equals(obj1 As Object, obj2 As Object) As Boolean
        If Not TypeOf obj1 Is Scala Then Return False
        Return DirectCast(obj1, Scala).Equals(obj2)
    End Function
#End Region

    Public Overrides Function ToString() As String
        Return Me.ToString(False)
    End Function

    Public Overloads Function ToString(bemollePrefer As Boolean) As String
        Return Me.NotaFondamentale.ToString(bemollePrefer)
    End Function

    Public Shared Function TryParse(value As String, ByRef result As Scala) As Boolean
        Dim nb As Nota = Nothing
        Dim modo As ModoBase = Nothing
        If value.Length > 3 AndAlso value.EndsWith("dim", StringComparison.CurrentCultureIgnoreCase) Then
            value = value.Substring(0, value.Length - 3)
        End If
        If Not Nota.TryParse(value.TrimEnd("m-+0123456789".ToCharArray), nb) Then Return False
        Dim regExNum As New System.Text.RegularExpressions.Regex("[0-9]+\+*")
        Dim accordoBase = regExNum.Replace(value, String.Empty)
        Dim na As Nota = Nothing
        Nota.TryParse(accordoBase, na)
        If nb.Equals(na) Then
            modo = ModoMaggiore.GetInstance
        Else
            modo = ModoMinoreArmonica.GetInstance
        End If
        result = New Scala(nb, modo)
        Return True
    End Function

End Class

File: Enzo.Music/Scala.cs
using System;
using System.Numerics;
using static System.Runtime.InteropServices.JavaScript.JSType;

namespace Enzo.Music;

public class Scala : IEqualityOperators<Scala, Scala, bool>
{
    public Nota NotaFondamentale { get; }

    public ModoBase _Modo;
    public ModoBase Modo => _Modo;

    public Scala(Nota notaFondamentale, ModoBase modo)
    {
        ArgumentNullException.ThrowIfNull(notaFondamentale, nameof(notaFondamentale));
        ArgumentNullException.ThrowIfNull(modo, nameof(modo));
        NotaFondamentale = notaFondamentale;
        _Modo = modo;
    }
    public Nota this[int index]
    {
        get
        {
            var incrementi = index % Modo.NumeroSuccessioni;
            Nota r = NotaFondamentale;
            for (int i = 0; i <= incrementi - 1; i++)
                if (!(Modo.Successioni?[i] is null)) r += Modo.Successioni[i];
            return r;
        }
    }

    #region Verifica uguaglianze
    public static bool operator ==(Scala? left, Scala? right)
    {
        if (left is null || right is null) return false;
        return left.Equals(right);
    }

    public static bool operator !=(Scala? left, Scala? right)
    {
        if (left is null || right is null) return false;
        return !left.Equals(right);
    }
    public override bool Equals(object? obj)
    {
        if (obj is null) return false;
        if (obj is not Scala) return false;
        Scala? s = obj as Scala;
        bool bNota = s?.NotaFondamentale?.Equals(NotaFondamentale) ?? false;
        bool bModo = s?.Modo?.Equals(s.Modo) ?? false;
        return bNota && bModo;
    }
    public static bool Equals(Scala? left, Scala? right)
    {
        if (left is null || right is null) return false;
        return left.Equals(right);
    }

    #endregion

    public override string ToString()
    {
        return ToString(false);
    }
    public string ToString(bool bemollePrefer)
    {
        return NotaFondamentale.ToString(bemollePrefer);
    }

    public override int GetHashCode()
    {
        return ToString().GetHashCode();
    }

    public static bool TryParse(string value, out Scala? result)
    {
        ModoBase? modo;
        if (value.Length > 3 && value.EndsWith("dim", StringComparison.CurrentCultureIgnoreCase))
            value = value[..^3];

        if (!Nota.TryParse(value.TrimEnd("m-+0123456789".ToCharArray()), out Nota? nb))
        {
            result = null;
            return false;
        }

        System.Text.RegularExpressions.Regex regExNum = new("[0-9]+\\+*");
        var accordoBase = regExNum.Replace(value, string.Empty);
        if (!Nota.TryParse(accordoBase, out Nota? na))
        {
            result = null;
            //return false;
        }

        if (Nota.Equals(nb, na))
            modo = ModoMaggiore.GetInstance();
        else
            modo = ModoMinoreArmonica.GetInstance();

        result = nb is null ? null : new Scala(nb, modo);
        return true;
    }
}
